### ISSUE TITLE
[Feature request] Adding labels and annotations to pods

### ISSUE THREAD

**Issue by jachinte** [AbdelrhmanHamouda/locust-k8s-operator#32]
**Is your feature request related to a problem? Please describe.**
I would like to get Prometheus metrics for tests I launch programmatically. Since Prometheus scrapes all pods annotated with `prometheus.io/scrape: true`, metrics are stored as if they belong to the same test / system.

**Describe the solution you'd like**
I'd like to attach labels and annotations to the deployed pods, so that I can customize the Prometheus config. This is what I have in mind: use `__meta_kubernetes_pod_label_<labelname>` and `__meta_kubernetes_pod_annotation_<annotationname>` in [kubernetes_sd_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config).

**Describe alternatives you've considered**
I am not aware of other solutions.

**Additional context**
I already implemented this feature in [looflow/locust-k8s-operator](https://github.com/looflow/locust-k8s-operator). Would you be interested in having this as a contribution?

**Comment by AbdelrhmanHamouda** [comment]
Hello @jachinte,
The project is open for contributions and all fresh perspectives  are more than welcome. The intent is to be as open and as flexible without sacrificing the core offerings of the tool.

**I would like to understand more about your use case.** I took a look at the implementation as well and I see that you wish to have the configuration of the labels on a _test CR_ level. 

To my understanding, the plan is to have `__meta_kubernetes_pod_label_<labelname>` and `__meta_kubernetes_pod_annotation_<annotationname> ` as part of your Prometheus **server** config which is usually a stable and not frequently updated part of the cloud infra. To my eyes, if the labels / annotations are changeable on a _test CR_ level, it means that those  labels / annotations  are subject to (sudden/per CR deployment) change -assuming no accidents- this would prompt Prometheus server config update + dashboard(s) query updates as well.  With that in mind, I am wondering if having the labels/annotations as part of the _Operator_ HELM config is a better placement. This will allow the labels to be more protected against "accidents", provide more governance capabilities to "cloud admins" & SREs and satisfy your request.  What do you think? 


That being said, if the purpose of this request is to mainly have a unique/predictable label that can be filtered and there is no other motivation for having the ability to influence injected labels / annotations , today the _Operator_ adds the following labels automatically to the generated resources for that purpose:
 - `performance-test-name`: Directly derived from the deployed custom resource name and is common between `master` and `worker` pods. The label can be used to filter and "group" metrics in the visualisation solution of your choice  i.e. Grafana,  NewRelic, etc...
   -  On master pods:  `performance-test-name=<cr.name>`
    - On worker pods:  `performance-test-name=<cr.name>`
- `performance-test-pod-name`: Also directly derived from the custom resource name and `-master`/`-worker` is appended to the end depending on the pod operation mode. This label can be used to further breakdown or further group the metrics 
  -  On master:  `performance-test-pod-name=<cr-name>-master`
  - On worker:  `performance-test-pod-name=<cr-name>-worker`

Currently the previously mentioned labels are used with Prometheus servers to scrap the data and visualize  test results in a NewRelic dashboard that breaks down all the metrics. 

what are your thoughts?

**Comment by jachinte** [comment]
Hello @AbdelrhmanHamouda,

Thank you for taking the time to analyze the feature I'm proposing, and for being this open to receive contributions. I hope this answer helps clarify my use case:

The pod labels and annotations provide unique identifiers related to the origin of the tests. Two labels I'm looking to include are the tenant id and test id. The annotations I have in mind will include parameters specific to the system under test for filtering purposes in Grafana.

I'm not expecting to change the Prometheus configuration often, but rather configure the monitoring and relabelling based on a set of known label and annotation keys. A Prometheus instance, plus other components, will use these labels to store the metrics appropriately (e.g., Thanos / Cortex will use the tenant id).

In my case, having the annotations or labels as part of the Helm chart is not a solution because that information is provided as part of the test CR. I think this approach would make sense when you have a namespace per client, thus having several instances of the locust operator.

I did notice the `performance-test-*` labels and annotations in the current implementation. However, my current use case requires adding custom / unique values, such as the ones I described above.

Here is an example:

```yaml
apiVersion: locust.io/v1
kind: LocustTest
metadata:
  name: demo.test
  namespace: locust-operator
spec:
  image: locustio/locust:latest
  labels:
    master:
      tenant: "XYZ999"
      test: "ABC123"
  annotations:
    master:
      param/users: "100"
  masterCommandSeed: >
    --locustfile /lotest/src/mysystem.py
    --host https://www.mysystem.com
    --users 100
    --spawn-rate 3
    --run-time 3m
  workerCommandSeed: --locustfile /lotest/src/mysystem.py
  workerReplicas: 3
  configMap: mysystem-cm
```

**Comment by AbdelrhmanHamouda** [comment]
Okay, I have slept on it and I think it is a good enhancement to add. Please feel free to open a PR.

Few remarks:
- I kindly invite you to take a look at the [CONTRIBUTING.md](https://www.amazon.es/gp/product/B08W2C5W59/ref=ox_sc_saved_title_2?smid=A291NOQ0IDRYUU&psc=1) if you haven't already. 
- In effort of making this project as maintainable as possible, please add tests to cover your added functionality. 
- I am planning to add an in-depth CRD / CR breakdown and its impact on generated resources to the documentation site. For that end, please also prepare a markdown doc to be included in one of the site pages. I'll provide more details during the PR review stage since the parent document is not yet ready.

**Comment by AbdelrhmanHamouda** [comment]
Request got implemented in #45
